File: dart_sample/lib/demos/material/range_slider_demo_page.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';

class RangeSliderDemoPage extends StatefulWidget {
  const RangeSliderDemoPage({super.key});

  @override
  State<RangeSliderDemoPage> createState() => _RangeSliderDemoPageState();
}

class _RangeSliderDemoPageState extends State<RangeSliderDemoPage> {
  bool _enabled = true;
  bool _discrete = false;
  bool _useThemeOverrides = false;
  bool _useWidgetColorOverride = false;
  bool _useMaterial3 = true;
  RangeValues _values = const RangeValues(0.2, 0.7);
  String _status = 'idle';

  @override
  Widget build(BuildContext context) {
    final ThemeData baseTheme = Theme.of(context);
    final ThemeData themedData = baseTheme.copyWith(
      useMaterial3: _useMaterial3,
      sliderTheme: _useThemeOverrides
          ? const SliderThemeData(
              activeTrackColor: Color(0xFF2E7D32),
              inactiveTrackColor: Color(0xFFC8E6C9),
              thumbColor: Color(0xFF1B5E20),
              disabledActiveTrackColor: Color(0x66212121),
              disabledInactiveTrackColor: Color(0x1F212121),
              disabledThumbColor: Color(0x66212121),
              trackHeight: 6,
            )
          : const SliderThemeData(),
    );

    return Theme(
      data: themedData,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: <Widget>[
          const Text(
            'RangeSlider baseline',
            style: TextStyle(fontSize: 20, color: Colors.black),
          ),
          const SizedBox(height: 8),
          const Text(
            'Two-thumb range mapping, continuous/discrete updates, drag/tap/keyboard flow, and M2/M3 theme/widget color precedence.',
            style: TextStyle(fontSize: 14, color: Color(0x8A000000)),
          ),
          const SizedBox(height: 10),
          Row(
            children: <Widget>[
              _buildControlButton(
                label: _enabled ? 'Enabled' : 'Disabled',
                onTap: () => setState(() => _enabled = !_enabled),
                width: 96,
                background: const Color(0xFFE9F0FF),
              ),
              const SizedBox(width: 8),
              _buildControlButton(
                label: _discrete ? 'Discrete' : 'Continuous',
                onTap: () => setState(() => _discrete = !_discrete),
                width: 112,
                background: const Color(0xFFE8F5E9),
              ),
              const SizedBox(width: 8),
              _buildControlButton(
                label: _useMaterial3 ? 'M3' : 'M2',
                onTap: () => setState(() => _useMaterial3 = !_useMaterial3),
                width: 76,
                background: const Color(0xFFFFF8E1),
              ),
              const SizedBox(width: 8),
              _buildControlButton(
                label: _useThemeOverrides ? 'Theme on' : 'Theme off',
                onTap: () =>
                    setState(() => _useThemeOverrides = !_useThemeOverrides),
                width: 112,
                background: const Color(0xFFEAF6F7),
              ),
            ],
          ),
          const SizedBox(height: 8),
          Row(
            children: <Widget>[
              _buildControlButton(
                label: _useWidgetColorOverride ? 'Widget on' : 'Widget off',
                onTap: () => setState(
                  () => _useWidgetColorOverride = !_useWidgetColorOverride,
                ),
                width: 118,
                background: const Color(0xFFF0E8FF),
              ),
              const SizedBox(width: 8),
              _buildControlButton(
                label: 'S-',
                onTap: () => _nudgeStart(-0.1),
                width: 42,
                background: const Color(0xFFFFF3E0),
              ),
              const SizedBox(width: 8),
              _buildControlButton(
                label: 'S+',
                onTap: () => _nudgeStart(0.1),
                width: 42,
                background: const Color(0xFFFFF3E0),
              ),
              const SizedBox(width: 8),
              _buildControlButton(
                label: 'E-',
                onTap: () => _nudgeEnd(-0.1),
                width: 42,
                background: const Color(0xFFFFF3E0),
              ),
              const SizedBox(width: 8),
              _buildControlButton(
                label: 'E+',
                onTap: () => _nudgeEnd(0.1),
                width: 42,
                background: const Color(0xFFFFF3E0),
              ),
              const SizedBox(width: 8),
              Expanded(
                child: Text(
                  'start=${_values.start.toStringAsFixed(2)}, end=${_values.end.toStringAsFixed(2)}, status=$_status',
                  style: const TextStyle(
                    fontSize: 12,
                    color: Color(0xFF607D8B),
                  ),
                ),
              ),
            ],
          ),
          const SizedBox(height: 10),
          Expanded(
            child: SingleChildScrollView(
              child: Column(
                children: <Widget>[
                  _buildPreviewCard(
                    title: 'LTR',
                    subtitle: 'Left-to-right mapping and keyboard direction',
                    textDirection: TextDirection.ltr,
                  ),
                  const SizedBox(height: 14),
                  _buildPreviewCard(
                    title: 'RTL',
                    subtitle: 'Right-to-left mapping and keyboard direction',
                    textDirection: TextDirection.rtl,
                  ),
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildPreviewCard({
    required String title,
    required String subtitle,
    required TextDirection textDirection,
  }) {
    return Container(
      color: const Color(0xFFF7F9FC),
      padding: const EdgeInsets.all(12),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: <Widget>[
          Text(
            title,
            style: const TextStyle(fontSize: 14, color: Colors.black),
          ),
          const SizedBox(height: 4),
          Text(
            subtitle,
            style: const TextStyle(fontSize: 12, color: Color(0x8A000000)),
          ),
          const SizedBox(height: 8),
          Directionality(
            textDirection: textDirection,
            child: _buildRangeSlider(),
          ),
        ],
      ),
    );
  }

  Widget _buildRangeSlider() {
    return RangeSlider(
      values: _values,
      min: 0,
      max: 1,
      divisions: _discrete ? 5 : null,
      activeColor: _useWidgetColorOverride ? const Color(0xFFB71C1C) : null,
      inactiveColor: _useWidgetColorOverride ? const Color(0xFFFFCDD2) : null,
      onChanged: _enabled ? _handleValuesChanged : null,
      onChangeStart: (RangeValues values) => setState(
        () =>
            _status = 'start ${values.start.toStringAsFixed(2)}-${values.end.toStringAsFixed(2)}',
      ),
      onChangeEnd: (RangeValues values) => setState(
        () =>
            _status = 'end ${values.start.toStringAsFixed(2)}-${values.end.toStringAsFixed(2)}',
      ),
      semanticFormatterCallback: (double value) => '${(value * 100).round()}%',
    );
  }

  void _handleValuesChanged(RangeValues values) {
    setState(() {
      _values = values;
      _status =
          'change ${values.start.toStringAsFixed(2)}-${values.end.toStringAsFixed(2)}';
    });
  }

  void _nudgeStart(double delta) {
    setState(() {
      final double nextStart = math.min(
        _values.end,
        math.max(0, _values.start + delta),
      );
      _values = RangeValues(nextStart, _values.end);
      _status =
          'manual ${_values.start.toStringAsFixed(2)}-${_values.end.toStringAsFixed(2)}';
    });
  }

  void _nudgeEnd(double delta) {
    setState(() {
      final double nextEnd = math.max(
        _values.start,
        math.min(1, _values.end + delta),
      );
      _values = RangeValues(_values.start, nextEnd);
      _status =
          'manual ${_values.start.toStringAsFixed(2)}-${_values.end.toStringAsFixed(2)}';
    });
  }

  Widget _buildControlButton({
    required String label,
    required VoidCallback onTap,
    required double width,
    required Color background,
  }) {
    return SizedBox(
      width: width,
      child: TextButton(
        onPressed: onTap,
        style: TextButton.styleFrom(
          backgroundColor: background,
          foregroundColor: Colors.black,
          minimumSize: const Size(0, 36),
          padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
          shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(8)),
        ),
        child: Text(label, style: const TextStyle(fontSize: 12)),
      ),
    );
  }
}

File: src/Sample/Plumix.Sample/Demos/Material/RangeSliderDemoPage.cs
using System;
using Avalonia;
using Avalonia.Media;
using Plumix.Material;
using Plumix.Rendering;
using Plumix.UI;
using Plumix.Widgets;

// Dart parity source (reference): dart_sample/lib/demos/material/range_slider_demo_page.dart (exact sample parity)

namespace Plumix;

public sealed class RangeSliderDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new RangeSliderDemoPageState();
    }
}

internal sealed class RangeSliderDemoPageState : State
{
    private bool _enabled = true;
    private bool _discrete;
    private bool _useThemeOverrides;
    private bool _useWidgetColorOverride;
    private bool _useMaterial3 = true;
    private bool _year2023 = true;
    private RangeValues _values = new(0.2, 0.7);
    private string _status = "idle";

    public override Widget Build(BuildContext context)
    {
        var baseTheme = Theme.Of(context);
        var themedData = baseTheme with
        {
            UseMaterial3 = _useMaterial3,
            SliderTheme = _useThemeOverrides
                ? new SliderThemeData(
                    ActiveTrackColor: Color.Parse("#FF2E7D32"),
                    InactiveTrackColor: Color.Parse("#FFC8E6C9"),
                    ThumbColor: Color.Parse("#FF1B5E20"),
                    DisabledActiveTrackColor: Color.Parse("#66212121"),
                    DisabledInactiveTrackColor: Color.Parse("#1F212121"),
                    DisabledThumbColor: Color.Parse("#66212121"),
                    TrackHeight: 6,
                    ThumbRadius: 11)
                : new SliderThemeData()
        };

        return new Theme(
            data: themedData,
            child: new Column(
                crossAxisAlignment: CrossAxisAlignment.Stretch,
                spacing: 10,
                children:
                [
                    new Text("RangeSlider baseline", fontSize: 20, color: Colors.Black),
                    new Text(
                        "Two-thumb range mapping, continuous/discrete updates, drag/tap plus Tab focus "
                        + "between the two thumbs, and M2/M3 theme/widget color precedence.",
                        fontSize: 14,
                        color: Color.Parse("#8A000000")),
                    new Row(
                        spacing: 8,
                        children:
                        [
                            BuildControlButton(
                                label: _enabled ? "Enabled" : "Disabled",
                                onTap: () => SetState(() => _enabled = !_enabled),
                                width: 96,
                                background: Color.Parse("#FFE9F0FF")),
                            BuildControlButton(
                                label: _discrete ? "Discrete" : "Continuous",
                                onTap: () => SetState(() => _discrete = !_discrete),
                                width: 112,
                                background: Color.Parse("#FFE8F5E9")),
                            BuildControlButton(
                                label: _useMaterial3 ? "M3" : "M2",
                                onTap: () => SetState(() => _useMaterial3 = !_useMaterial3),
                                width: 76,
                                background: Color.Parse("#FFFFF8E1")),
                            BuildControlButton(
                                label: _useThemeOverrides ? "Theme on" : "Theme off",
                                onTap: () => SetState(() => _useThemeOverrides = !_useThemeOverrides),
                                width: 112,
                                background: Color.Parse("#FFEAF6F7")),
                        ]),
                    new Row(
                        spacing: 8,
                        children:
                        [
                            BuildControlButton(
                                label: _useWidgetColorOverride ? "Widget on" : "Widget off",
                                onTap: () => SetState(() => _useWidgetColorOverride = !_useWidgetColorOverride),
                                width: 118,
                                background: Color.Parse("#FFF0E8FF")),
                            BuildControlButton(
                                label: "S-",
                                onTap: () => NudgeStart(-0.1),
                                width: 42,
                                background: Color.Parse("#FFFFF3E0")),
                            BuildControlButton(
                                label: "S+",
                                onTap: () => NudgeStart(0.1),
                                width: 42,
                                background: Color.Parse("#FFFFF3E0")),
                            BuildControlButton(
                                label: "E-",
                                onTap: () => NudgeEnd(-0.1),
                                width: 42,
                                background: Color.Parse("#FFFFF3E0")),
                            BuildControlButton(
                                label: "E+",
                                onTap: () => NudgeEnd(0.1),
                                width: 42,
                                background: Color.Parse("#FFFFF3E0")),
                            new Expanded(
                                child: new Text(
                                    $"start={_values.Start:0.00}, end={_values.End:0.00}, status={_status}",
                                    fontSize: 12,
                                    color: Color.Parse("#FF607D8B"))),
                        ]),
                    new Row(
                        children:
                        [
                            BuildControlButton(
                                label: _year2023 ? "2023 look" : "2024 look",
                                onTap: () => SetState(() => _year2023 = !_year2023),
                                width: 96,
                                background: Color.Parse("#FFEAF6F7")),
                        ]),
                    new Expanded(
                        child: new SingleChildScrollView(
                            child: new Column(
                                crossAxisAlignment: CrossAxisAlignment.Stretch,
                                spacing: 14,
                                children:
                                [
                                    BuildPreviewCard(
                                        title: "LTR",
                                        subtitle: "Left-to-right mapping and thumb order",
                                        textDirection: TextDirection.Ltr),
                                    BuildPreviewCard(
                                        title: "RTL",
                                        subtitle: "Right-to-left mapping and thumb order",
                                        textDirection: TextDirection.Rtl),
                                ]))),
                ]));
    }

    private Widget BuildPreviewCard(string title, string subtitle, TextDirection textDirection)
    {
        return new Container(
            color: Color.Parse("#FFF7F9FC"),
            padding: new Thickness(12),
            child: new Column(
                crossAxisAlignment: CrossAxisAlignment.Stretch,
                spacing: 8,
                children:
                [
                    new Text(title, fontSize: 14, color: Colors.Black),
                    new Text(subtitle, fontSize: 12, color: Color.Parse("#8A000000")),
                    new Directionality(
                        textDirection: textDirection,
                        child: BuildRangeSlider()),
                ]));
    }

    private Widget BuildRangeSlider()
    {
        return new RangeSlider(
            values: _values,
            min: 0,
            max: 1,
            divisions: _discrete ? 5 : null,
            labels: new RangeLabels(
                $"{Math.Round(_values.Start * 100)}",
                $"{Math.Round(_values.End * 100)}"),
            activeColor: _useWidgetColorOverride ? Color.Parse("#FFB71C1C") : null,
            inactiveColor: _useWidgetColorOverride ? Color.Parse("#FFFFCDD2") : null,
            year2023: _year2023,
            onChanged: _enabled ? HandleValuesChanged : null,
            onChangeStart: values => SetState(() => _status = $"start {values.Start:0.00}-{values.End:0.00}"),
            onChangeEnd: values => SetState(() => _status = $"end {values.Start:0.00}-{values.End:0.00}"),
            semanticFormatterCallback: value => $"{Math.Round(value * 100):0}%");
    }

    private void HandleValuesChanged(RangeValues values)
    {
        SetState(() =>
        {
            _values = values;
            _status = $"change {values.Start:0.00}-{values.End:0.00}";
        });
    }

    private void NudgeStart(double delta)
    {
        SetState(() =>
        {
            double nextStart = Math.Clamp(_values.Start + delta, 0, _values.End);
            _values = new RangeValues(nextStart, _values.End);
            _status = $"manual {nextStart:0.00}-{_values.End:0.00}";
        });
    }

    private void NudgeEnd(double delta)
    {
        SetState(() =>
        {
            double nextEnd = Math.Clamp(_values.End + delta, _values.Start, 1);
            _values = new RangeValues(_values.Start, nextEnd);
            _status = $"manual {_values.Start:0.00}-{nextEnd:0.00}";
        });
    }

    private Widget BuildControlButton(
        string label,
        Action onTap,
        double width,
        Color background)
    {
        return new SizedBox(
            width: width,
            child: new TextButton(
                onPressed: onTap,
                backgroundColor: background,
                foregroundColor: Colors.Black,
                minHeight: 36,
                padding: new Thickness(10, 8),
                borderRadius: BorderRadius.Circular(8),
                child: new Text(label, fontSize: 12)));
    }
}
